### ISSUE TITLE
Charm errorring with `plugin configuration secret was not provided` even though secret-id is set

### ISSUE THREAD

**Issue by shipperizer** [canonical/lego-operator#17]
### Bug Description

charm is provided a secret id via config but charm is not picking that up

### To Reproduce

used the following terraform in our private cloud, used vault for providing the secret

```
terraform {
  required_version = ">= 1.6.6"
  required_providers {
    juju = {
      source  = "juju/juju"
      version = ">= 0.15.0"
    }
    openstack = {
      source  = "terraform-provider-openstack/openstack"
      version = ">= 1.53.0"
    }

    vault = {
      source  = "hashicorp/vault"
      version = ">= 3.23.0"
    }
  }
}


variable "model" {
  description = "The name of the Juju model."
  type        = string
  default     = ""
}

variable "name" {
  description = "The name of the Juju application."
  type        = string
  default     = "lego"
}

variable "trust" {
  description = "The status to grant the Juju application full access to the cluster."
  type        = bool
  default     = true
}

variable "secret_path" {
  description = "The path of the secret containing the charm sensitive configuration."
  type        = string
}


variable "charm" {
  description = "The application charm operator information."
  type = object({
    name : string
    channel : string
    base : string
  })
  default = {
    name    = "lego"
    channel = "4/edge"
    base    = "ubuntu@22.04"
  }
}

variable "lego_config" {
  type = object({
    email                       = optional(string, "email@test.com")
    httpreq_endpoint            = optional(string, "https://lego-certs.test.com")
    httpreq_http_timeout        = optional(number, 180)
    httpreq_propagation_timeout = optional(number, 600)
  })
  description = <<EOT
    jimm_config = {
      email: "Account email address, this email will receive notifications from Let's Encrypt"
      httpreq_endpoint: "URL of the custom API that can be used to fulfill the DNS-01 challenge."
      httpreq_http_timeout: "API request timeout (seconds)."
      httpreq_propagation_timeout: "Maximum waiting time for DNS propagation (seconds)"
    }
  EOT
  default = {
    email                       = "email@test.com")
    httpreq_endpoint            = "https://lego-certs.test.com"
    httpreq_http_timeout        = 180
    httpreq_propagation_timeout = 600
  }
}

data "juju_model" "model" {
  name = var.model
}


data "vault_generic_secret" "lego_credentials" {
  path = var.secret_path
}


resource "juju_secret" "lego_plugin_secret" {
  model = data.juju_model.model.name
  name  = "lego-httpreq-plugin-secret"
  value = {
    httpreq-endpoint            = var.lego_config.httpreq_endpoint
    httpreq-http_timeout        = var.lego_config.httpreq_http_timeout
    httpreq-propagation_timeout = var.lego_config.httpreq_propagation_timeout
    httpreq-username            = data.vault_generic_secret.lego_credentials.data["username"]
    httpreq-password            = sensitive(data.vault_generic_secret.lego_credentials.data["password"])
  }
  info = "Secret for LEGO charm, httpreq config information included"
}

resource "juju_access_secret" "grant" {
  secret_id    = juju_secret.lego_plugin_secret.secret_id
  model        = data.juju_model.model.name
  applications = [juju_application.lego.name]
}

resource "juju_application" "lego" {
  name  = var.name
  model = data.juju_model.model.name
  trust = var.trust

  charm {
    name    = var.charm.name
    channel = var.charm.channel
    base    = var.charm.base
  }

  config = {
    email                   = var.lego_config.email
    plugin                  = "httpreq"
    plugin-config-secret-id = juju_secret.lego_plugin_secret.secret_id
  }
}
```

where the vault secret was 

```
me@box:~$ vault kv get secret/dev-core/lego
====== Data ======
Key         Value
---         -----
password    **************************
username    something

```

and the juju secret is 

```
me@box:~$ juju show-secret ctc1u57mp25c77lu5tsg --reveal
ctc1u57mp25c77lu5tsg:
  revision: 2
  owner: <model>
  description: Secret for LEGO charm, httpreq config information included
  name: lego-httpreq-plugin-secret
  created: 2024-12-10T10:59:33Z
  updated: 2024-12-10T11:06:43Z
  content:
    httpreq-endpoint: https://lego-certs.test.com
    httpreq-http-timeout: "180"
    httpreq-password: ********************
    httpreq-propagation-timeout: "600"
    httpreq-username: something
  access:
  - target: application-lego
    scope: model-f931006d-c35c-440e-879e-622169f4eb7b
    role: view

```

### Environment

`4/edge` and `4/beta`

### Relevant log output

```shell
ubuntu@dev-k8s-leader:~$ sudo k8s kubectl logs -n dev-core lego-0 -f
Defaulted container "charm" out of: charm, charm-init (init)
2024-12-10T11:43:07.651Z [pebble] HTTP API server listening on ":38812".
2024-12-10T11:43:07.652Z [pebble] Started daemon.
2024-12-10T11:43:07.657Z [pebble] POST /v1/services 2.757764ms 202
2024-12-10T11:43:07.659Z [pebble] Service "container-agent" starting: /charm/bin/containeragent unit --data-dir /var/lib/juju --append-env "PATH=$PATH:/charm/bin" --show-log --charm-modified-version 1
2024-12-10T11:43:07.688Z [container-agent] 2024-12-10 11:43:07 INFO juju.cmd supercommand.go:56 running containerAgent [3.5.5 1d21840563a55809580976df0b4880e03323bbb7 gc go1.23.3]
2024-12-10T11:43:07.688Z [container-agent] starting containeragent unit command
2024-12-10T11:43:07.688Z [container-agent] containeragent unit "unit-lego-0" start (3.5.5 [gc])
2024-12-10T11:43:07.688Z [container-agent] 2024-12-10 11:43:07 INFO juju.cmd.containeragent.unit runner.go:592 start "unit"
2024-12-10T11:43:07.688Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.upgradesteps worker.go:60 upgrade steps for 3.5.5 have already been run.
2024-12-10T11:43:07.690Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.probehttpserver server.go:157 starting http server on 127.0.0.1:65301
2024-12-10T11:43:07.713Z [container-agent] 2024-12-10 11:43:07 INFO juju.api apiclient.go:571 connection established to "wss://controller-service.controller-k8s-dev.svc.cluster.local:17070/model/f931006d-c35c-440e-879e-622169f4eb7b/api"
2024-12-10T11:43:07.717Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.apicaller connect.go:163 [f93100] "unit-lego-0" successfully connected to "controller-service.controller-k8s-dev.svc.cluster.local:17070"
2024-12-10T11:43:07.734Z [container-agent] 2024-12-10 11:43:07 INFO juju.api apiclient.go:571 connection established to "wss://10.142.140.117:17070/model/f931006d-c35c-440e-879e-622169f4eb7b/api"
2024-12-10T11:43:07.737Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.apicaller connect.go:163 [f93100] "unit-lego-0" successfully connected to "10.142.140.117:17070"
2024-12-10T11:43:07.764Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.migrationminion worker.go:147 migration phase is now: NONE
2024-12-10T11:43:07.765Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.logger logger.go:120 logger worker started
2024-12-10T11:43:07.779Z [container-agent] 2024-12-10 11:43:07 WARNING juju.worker.proxyupdater proxyupdater.go:241 unable to set snap core settings [proxy.http= proxy.https= proxy.store=]: exec: "snap": executable file not found in $PATH, output: ""
2024-12-10T11:43:07.786Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.leadership tracker.go:194 lego/0 promoted to leadership of lego
2024-12-10T11:43:07.789Z [container-agent] 2024-12-10 11:43:07 INFO juju.agent.tools symlinks.go:20 ensure jujuc symlinks in /var/lib/juju/tools/unit-lego-0
2024-12-10T11:43:07.791Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.caasupgrader upgrader.go:113 abort check blocked until version event received
2024-12-10T11:43:07.791Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.caasupgrader upgrader.go:119 unblocking abort check
2024-12-10T11:43:07.847Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.uniter uniter.go:363 unit "lego/0" started
2024-12-10T11:43:07.858Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.uniter uniter.go:389 hooks are retried true
2024-12-10T11:43:07.953Z [container-agent] 2024-12-10 11:43:07 INFO juju.worker.uniter.charm bundles.go:82 downloading ch:amd64/jammy/lego-15 from API server
2024-12-10T11:43:07.953Z [container-agent] 2024-12-10 11:43:07 INFO juju.downloader download.go:112 downloading from ch:amd64/jammy/lego-15
2024-12-10T11:43:08.671Z [pebble] GET /v1/changes/1/wait 1.013143s 200
2024-12-10T11:43:08.671Z [pebble] Started default services with change 1.
2024-12-10T11:43:08.831Z [container-agent] 2024-12-10 11:43:08 INFO juju.downloader download.go:95 download complete ("ch:amd64/jammy/lego-15")
2024-12-10T11:43:08.884Z [container-agent] 2024-12-10 11:43:08 INFO juju.downloader download.go:175 download verified ("ch:amd64/jammy/lego-15")
2024-12-10T11:43:13.999Z [container-agent] 2024-12-10 11:43:13 INFO juju.worker.uniter resolver.go:185 found queued "upgrade-charm" hook
2024-12-10T11:43:14.647Z [container-agent] 2024-12-10 11:43:14 INFO juju-log Running legacy hooks/upgrade-charm.
2024-12-10T11:43:15.527Z [container-agent] 2024-12-10 11:43:15 INFO juju.worker.uniter.operation runhook.go:186 ran "upgrade-charm" hook (via hook dispatching script: dispatch)
2024-12-10T11:43:15.542Z [container-agent] 2024-12-10 11:43:15 INFO juju.worker.uniter resolver.go:185 found queued "config-changed" hook
2024-12-10T11:43:16.137Z [container-agent] 2024-12-10 11:43:16 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T11:43:16.643Z [container-agent] 2024-12-10 11:43:16 INFO juju.worker.uniter.operation runhook.go:186 ran "config-changed" hook (via hook dispatching script: dispatch)
2024-12-10T11:43:16.677Z [container-agent] 2024-12-10 11:43:16 INFO juju.worker.uniter resolver.go:76 reboot detected; triggering implicit start hook to notify charm
2024-12-10T11:43:17.034Z [container-agent] 2024-12-10 11:43:17 INFO juju-log Running legacy hooks/start.
2024-12-10T11:43:17.654Z [pebble] Check "readiness" failure 1 (threshold 3): received non-20x status code 418
2024-12-10T11:43:17.810Z [container-agent] 2024-12-10 11:43:17 INFO juju.worker.uniter.operation runhook.go:186 ran "start" hook (via hook dispatching script: dispatch)
2024-12-10T11:48:23.176Z [container-agent] 2024-12-10 11:48:23 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T11:48:23.509Z [container-agent] 2024-12-10 11:48:23 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T11:54:21.143Z [container-agent] 2024-12-10 11:54:21 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T11:54:21.502Z [container-agent] 2024-12-10 11:54:21 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T11:59:38.656Z [container-agent] 2024-12-10 11:59:38 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T11:59:38.994Z [container-agent] 2024-12-10 11:59:38 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:04:37.580Z [container-agent] 2024-12-10 12:04:37 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:04:37.923Z [container-agent] 2024-12-10 12:04:37 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:10:06.140Z [container-agent] 2024-12-10 12:10:06 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:10:06.470Z [container-agent] 2024-12-10 12:10:06 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:15:57.120Z [container-agent] 2024-12-10 12:15:57 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:15:57.522Z [container-agent] 2024-12-10 12:15:57 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:21:55.549Z [container-agent] 2024-12-10 12:21:55 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:21:55.922Z [container-agent] 2024-12-10 12:21:55 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:26:58.687Z [container-agent] 2024-12-10 12:26:58 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:26:59.016Z [container-agent] 2024-12-10 12:26:59 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:32:33.876Z [container-agent] 2024-12-10 12:32:33 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:32:34.210Z [container-agent] 2024-12-10 12:32:34 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:38:32.186Z [container-agent] 2024-12-10 12:38:32 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:38:32.581Z [container-agent] 2024-12-10 12:38:32 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:44:16.473Z [container-agent] 2024-12-10 12:44:16 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:44:16.809Z [container-agent] 2024-12-10 12:44:16 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:49:39.866Z [container-agent] 2024-12-10 12:49:39 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:49:40.225Z [container-agent] 2024-12-10 12:49:40 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T12:54:50.007Z [container-agent] 2024-12-10 12:54:50 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T12:54:50.413Z [container-agent] 2024-12-10 12:54:50 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:00:06.509Z [container-agent] 2024-12-10 13:00:06 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:00:06.847Z [container-agent] 2024-12-10 13:00:06 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:04:13.372Z [container-agent] 2024-12-10 13:04:13 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:04:13.711Z [container-agent] 2024-12-10 13:04:13 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:08:36.099Z [container-agent] 2024-12-10 13:08:36 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:08:36.427Z [container-agent] 2024-12-10 13:08:36 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:13:10.138Z [container-agent] 2024-12-10 13:13:10 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:13:10.477Z [container-agent] 2024-12-10 13:13:10 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:17:13.641Z [container-agent] 2024-12-10 13:17:13 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:17:13.989Z [container-agent] 2024-12-10 13:17:13 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:21:14.515Z [container-agent] 2024-12-10 13:21:14 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:21:14.850Z [container-agent] 2024-12-10 13:21:14 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:25:42.662Z [container-agent] 2024-12-10 13:25:42 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:25:43.005Z [container-agent] 2024-12-10 13:25:43 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)
2024-12-10T13:31:00.639Z [container-agent] 2024-12-10 13:31:00 ERROR juju-log charm config validation failed: plugin configuration secret was not provided
2024-12-10T13:31:01.038Z [container-agent] 2024-12-10 13:31:01 INFO juju.worker.uniter.operation runhook.go:186 ran "update-status" hook (via hook dispatching script: dispatch)

```

### Additional context

![image](https://github.com/user-attachments/assets/e6e039d7-bdcf-42a1-81fe-0b34c31c91e8)
![image](https://github.com/user-attachments/assets/bdbec7f9-7f4e-431f-aa8d-30e99f0298db)

**Comment by shipperizer** [comment]
```
ubuntu@dev-k8s-leader:~$ juju config lego plugin-config-secret-id
ctc1u57mp25c77lu5tsg

```

**Comment by DanielArndt** [comment]
Thanks. I see now that was included in one of the images but it was a little hard to parse.

This is interesting. that error message comes from a single place:

https://github.com/canonical/lego-operator/blob/5428ea9e4f39927ea482f022757f804d015164d3/src/charm.py#L86-L88

Which if we follow it, means that `_plugin_config` returned a falsey value.

There's only a few cases I can see:

We haven't set the config value correctly, which is what I was trying to rule out.

https://github.com/canonical/lego-operator/blob/5428ea9e4f39927ea482f022757f804d015164d3/src/charm.py#L253-L255

We don't have permission to access the secret, or it isn't there. It looks like the `juju show-secret` command rules that out.

https://github.com/canonical/lego-operator/blob/5428ea9e4f39927ea482f022757f804d015164d3/src/charm.py#L256-L259

Or there's no values in the secret, which again, the `show-secret` results seem to rule out.

https://github.com/canonical/lego-operator/blob/5428ea9e4f39927ea482f022757f804d015164d3/src/charm.py#L260

I'm going to try to reproduce something, but I'm not sure how successful I'll be. We may need to add more logging (which we likely should have anyway) to get to the bottom of this.

**Comment by DanielArndt** [comment]
@shipperizer Can you try with the latest `4/edge` (which has #18) and let me know what the warning log says, assuming it shows up?

**Comment by DanielArndt** [comment]
Closing the loop from a conversation on Matrix:

> Alessandro Cabbia: i ve just run rev24 and somehow now it s working fine
only thing i tweaked was the proxy settings

So, I'll close this.

@shipperizer Please re-open this if it happens again!
